add spec for feemarket module (#889)
tharsis/ethermint: x/feemarket/spec/01_concepts.md

<!--
order: 1
-->

> # Concepts

Reviewer: Add a section on the `EIP 1559 Fee Market` to give context. I pulled up some of my previous notes, maybe it's helpful: 

- Fixed-per-block network fee (`baseFee`) that is burned
  - `baseFee` per gas changes according to a formula which is a function of gas used in parent block and gas target (block gas limit divided by elasticity multiplier) of parent block.
  - increases when blocks are above the gas target
  - decreases when blocks are below the gas target
- Dynamically expands/contracts block sizes to deal with transient congestion.
- Transactions specify a maximum fee per gas they are willing to pay total (aka: max fee), which covers both the priority fee and the block's network fee per gas (aka: base fee)

## Base fee

The base fee is a global base fee defined at the consensus level. It is adjusted for each block based on the total gas used in the previous block.

> Unlike the Cosmos SDK local `minimal-gas-prices`, this value is persisted in the KVStore which make a reliable value for validators to agree upon.

Reviewer: Unlike the Cosmos SDK local `minimal-gas-prices`, this value is persisted in the KVStore which provides a reliable value for validators to agree upon.

<!--
order: 1
-->

# Concepts

## Base fee

The base fee is a global base fee defined at the consensus level. It is adjusted for each block based on the total gas used in the previous block and gas target (block gas limit divided by elasticity multiplier) 

- it increases when blocks are above the gas target
- it decreases when blocks are below the gas target

Unlike the Cosmos SDK local `minimal-gas-prices`, this value is stored as a module parameter which provides a reliable value for validators to agree upon.

## Tip

In EIP-1559, the `tip` is a value that can be added to the `baseFee` in order to incentive transaction prioritization.

The transaction fee in Ethereum is calculated using the following the formula :

`transaction fee = (baseFee + tip) * gas units (limit)`

> In Cosmos SDK there is no notion of prioritization, thus the tip for an EIP-1559 transactions in Ethermint should be zero (MaxPriorityFeePerGas endpoint returns 0)

Author: ```suggestion
In Cosmos SDK there is no notion of prioritization, thus the tip for an EIP-1559 transaction in Ethermint should be zero (MaxPriorityFeePerGas json-rpc endpoint returns 0)
```

> In Cosmos SDK there is no notion of prioritization, thus the tip for an EIP-1559 transaction in Ethermint should be zero (MaxPriorityFeePerGas json-rpc endpoint returns 0)

Reviewer 2: ```suggestion
In Cosmos SDK there is no notion of prioritization, thus the tip for an EIP-1559 transaction in Ethermint should be zero (`MaxPriorityFeePerGas` JSON-RPC endpoint returns `0`)
```

## EIP-1559

A transaction pricing mechanism introduced in Ethereum that includes fixed-per-block network fee that is burned and dynamically expands/contracts block sizes to deal with transient congestion.

Transactions specify a maximum fee per gas they are willing to pay total (aka: max fee), which covers both the priority fee and the block's network fee per gas (aka: base fee)

> Reference:
> https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md

Reviewer 2: ```suggestion
Reference: [EIP1559](https://eips.ethereum.org/EIPS/eip-1559)
```
